File: src/main/java/net/gegy1000/terrarium/server/world/generator/EarthGenerationHandler.java
package net.gegy1000.terrarium.server.world.generator;

import net.gegy1000.terrarium.Terrarium;
import net.gegy1000.terrarium.server.capability.TerrariumWorldData;
import net.gegy1000.terrarium.server.map.GenerationRegion;
import net.gegy1000.terrarium.server.map.GenerationRegionHandler;
import net.gegy1000.terrarium.server.map.cover.CoverType;
import net.gegy1000.terrarium.server.map.source.glob.CoverTileAccess;
import net.gegy1000.terrarium.server.map.source.height.HeightTileAccess;
import net.gegy1000.terrarium.server.map.system.component.TerrariumComponentTypes;
import net.gegy1000.terrarium.server.world.EarthGenerationSettings;
import net.minecraft.util.math.MathHelper;

import java.util.Random;

public class EarthGenerationHandler {
    private final long worldSeed;
    private final EarthGenerationSettings settings;

    private final GenerationRegionHandler regionHandler;

    private final int maxHeight;

    private final int oceanHeight;

    private final Random random = new Random();

    public EarthGenerationHandler(long worldSeed, TerrariumWorldData worldData, EarthGenerationSettings settings, int maxHeight) {
        this.worldSeed = worldSeed;
        this.settings = settings;
        this.maxHeight = maxHeight;

        this.regionHandler = new GenerationRegionHandler(worldData, this);

        this.oceanHeight = this.settings.heightOffset + 1;
    }

    private void initializeSeed(int chunkX, int chunkZ) {
        long seed = chunkX * 132897987541L + chunkZ * 341873128712L;
        this.random.setSeed(seed ^ this.worldSeed);
    }

    public void populateHeightRegion(int[] buffer, int chunkX, int chunkZ) {
        this.initializeSeed(chunkX, chunkZ);

        int globalX = chunkX << 4;
        int globalZ = chunkZ << 4;

        try {
            for (int localZ = 0; localZ < 16; localZ++) {
                int blockZ = globalZ + localZ;

                for (int localX = 0; localX < 16; localX++) {
                    int blockX = globalX + localX;

                    GenerationRegion region = this.regionHandler.get(blockX, blockZ);
                    HeightTileAccess heightTile = region.getData().get(TerrariumComponentTypes.HEIGHT);

                    if (heightTile != null) {
                        short height = heightTile.getShort(blockX - region.getMinX(), blockZ - region.getMinZ());
                        int offsetHeight = height + this.settings.heightOffset;
                        buffer[localX + localZ * 16] = MathHelper.clamp(offsetHeight, 0, this.maxHeight);
                    }
                }
            }
        } catch (Exception e) {
            Terrarium.LOGGER.error("Failed to populate heightmap for {}, {}", chunkX, chunkZ, e);
        }
    }

    public void populateCoverRegion(CoverType[] buffer, int chunkX, int chunkZ) {
        this.initializeSeed(chunkX, chunkZ);

        int globalX = chunkX << 4;
        int globalZ = chunkZ << 4;

        try {
            for (int localZ = 0; localZ < 16; localZ++) {
                int blockZ = globalZ + localZ;
                for (int localX = 0; localX < 16; localX++) {
                    int blockX = globalX + localX;
                    buffer[localX + localZ * 16] = this.getCoverScattered(blockX, blockZ);
                }
            }
        } catch (Exception e) {
            Terrarium.LOGGER.error("Failed to populate cover region for {}, {}", chunkX, chunkZ, e);
        }
    }

    public void populateCoverDirect(CoverType[] buffer, int globalX, int globalZ, int width, int height) {
        if (buffer.length != width * height) {
            throw new IllegalArgumentException("Given width and height that do not match given buffer size");
        }

        try {
            for (int localZ = 0; localZ < height; localZ++) {
                int blockZ = globalZ + localZ;
                for (int localX = 0; localX < width; localX++) {
                    int blockX = globalX + localX;
                    buffer[localX + localZ * width] = this.getCover(blockX, blockZ);
                }
            }
        } catch (Exception e) {
            Terrarium.LOGGER.error("Failed to populate cover region at {}, {}", globalX, globalZ, e);
        }
    }

    private CoverType getCoverScattered(int x, int z) {
        CoverType originCover = this.getCover(x, z);

        int range = Math.max(1, MathHelper.ceil(this.settings.scatterRange * originCover.getScatterRange()));

        int scatterX = x + this.random.nextInt(range) - this.random.nextInt(range);
        int scatterZ = z + this.random.nextInt(range) - this.random.nextInt(range);

        CoverType scattered = this.getCover(scatterX, scatterZ);

        if (!scattered.canScatterTo()) {
            return originCover;
        }

        return scattered;
    }

    private CoverType getCover(int x, int z) {
        GenerationRegion region = this.regionHandler.get(x, z);
        CoverTileAccess coverTile = region.getData().get(TerrariumComponentTypes.COVER);

        if (coverTile != null) {
            return coverTile.get(x - region.getMinX(), z - region.getMinZ());
        }

        return CoverType.NO_DATA;
    }

    public EarthGenerationSettings getSettings() {
        return this.settings;
    }

    public int getOceanHeight() {
        return this.oceanHeight;
    }
}

File: src/main/kotlin/net/gegy1000/terrarium/server/world/generator/EarthGenerationHandler.kt
package net.gegy1000.terrarium.server.world.generator

import net.gegy1000.terrarium.Terrarium
import net.gegy1000.terrarium.server.capability.TerrariumWorldData
import net.gegy1000.terrarium.server.map.GenerationRegionHandler
import net.gegy1000.terrarium.server.map.RegionTilePos
import net.gegy1000.terrarium.server.map.glob.GlobType
import net.gegy1000.terrarium.server.util.Interpolation
import net.gegy1000.terrarium.server.world.EarthGenerationSettings
import net.minecraft.util.math.MathHelper
import java.util.Random

class EarthGenerationHandler(worldData: TerrariumWorldData, val settings: EarthGenerationSettings, val maxHeight: Int) {
    companion object {
        const val DATA_WIDTH = 432000
        const val DATA_HEIGHT = 216000
        const val REAL_SCALE = 92.766203
    }

    val random = Random()

    val oceanHeight = this.settings.heightOffset + 1
    val scatterRange = (this.settings.scatterRange * this.settings.scale).toInt()

    val finalScale = this.settings.scale * REAL_SCALE
    val heightScale = this.settings.scale * this.settings.terrainHeightScale

    val regionHandler = GenerationRegionHandler(worldData, this)

    fun initializeSeed(pos: RegionTilePos) {
        random.setSeed(pos.tileX.toLong() * 341873128712L + pos.tileZ.toLong() * 132897987541L)
    }

    fun scaleHeightRegion(result: ShortArray, sample: ShortArray, width: Int, height: Int, scaledWidth: Int, scaledHeight: Int) {
        for (localZ in 0 until scaledHeight) {
            val scaledZ = localZ * settings.scaleRatioZ
            val originZ = MathHelper.floor(scaledZ)
            val intermediateZ = scaledZ - originZ

            for (localX in 0 until scaledWidth) {
                val scaledX = localX * settings.scaleRatioX
                val originX = MathHelper.floor(scaledX)
                val intermediateX = scaledX - originX

                val sampleIndex = originX + originZ * width
                val current = sample[sampleIndex].toDouble()
                val south = sample[sampleIndex + width].toDouble()
                val east = sample[sampleIndex + 1].toDouble()
                val southEast = sample[sampleIndex + width + 1].toDouble()

                val y1 = Interpolation.cosine(current, south, intermediateZ)
                val y2 = Interpolation.cosine(east, southEast, intermediateZ)

                val interpolated = Interpolation.cosine(y1, y2, intermediateX)
                val scaled = (interpolated * this.heightScale).toInt()

                val resultIndex = localX + localZ * scaledWidth
                if (interpolated >= 0.0 && scaled < 1) {
                    result[resultIndex] = (this.settings.heightOffset + 1).toShort()
                } else {
                    result[resultIndex] = MathHelper.clamp(scaled + this.settings.heightOffset, 0, this.maxHeight).toShort()
                }
            }
        }
    }

    fun scaleGlobRegion(result: Array<GlobType>, sample: Array<GlobType>, width: Int, height: Int, scaledWidth: Int, scaledHeight: Int) {
        for (localZ in 0 until scaledHeight) {
            val scaledZ = localZ * settings.scaleRatioZ
            val originZ = MathHelper.floor(scaledZ)

            for (localX in 0 until scaledWidth) {
                val scaledX = localX * settings.scaleRatioX
                val originX = MathHelper.floor(scaledX)

                result[localX + localZ * scaledWidth] = sample[originX + originZ * width]
            }
        }
    }

    fun populateHeightRegion(buffer: IntArray, chunkX: Int, chunkZ: Int) {
        try {
            val x = chunkX shl 4
            val z = chunkZ shl 4

            for (localZ in 0..15) {
                val blockZ = z + localZ
                val globalOriginZ = MathHelper.floor(blockZ * settings.scaleRatioZ)

                for (localX in 0..15) {
                    val blockX = x + localX
                    val globalOriginX = MathHelper.floor(blockX * settings.scaleRatioX)

                    val region = this.regionHandler[globalOriginX, globalOriginZ]
                    buffer[localX + localZ * 16] = region.getHeight(blockX, blockZ)
                }
            }
        } catch (e: Exception) {
            buffer.fill(5)
            Terrarium.LOGGER.error("Failed to generate heightmap for $chunkX, $chunkZ", e)
        }
    }

    fun populateGlobRegion(buffer: Array<GlobType>, chunkX: Int, chunkZ: Int) {
        try {
            val x = chunkX shl 4
            val z = chunkZ shl 4

            for (localZ in 0..15) {
                val blockZ = z + localZ
                for (localX in 0..15) {
                    val blockX = x + localX
                    buffer[localX + localZ * 16] = getGlobScattered(blockX, blockZ)
                }
            }
        } catch (e: Exception) {
            buffer.fill(GlobType.NO_DATA)
            Terrarium.LOGGER.error("Failed to generate globcover map for $chunkX, $chunkZ", e)
        }
    }

    private fun getGlobScattered(x: Int, z: Int): GlobType {
        val originGlob = this.getGlob(x, z)

        val range = MathHelper.ceil(this.scatterRange * originGlob.scatterScale)

        val scatterX = x + this.random.nextInt(range) - this.random.nextInt(range)
        val scatterZ = z + this.random.nextInt(range) - this.random.nextInt(range)

        val scattered = this.getGlob(scatterX, scatterZ)

        if (!scattered.scatterTo) {
            return originGlob
        }

        return scattered
    }

    private fun getGlob(x: Int, z: Int): GlobType {
        val globalOriginX = MathHelper.floor(x * settings.scaleRatioX)
        val globalOriginZ = MathHelper.floor(z * settings.scaleRatioZ)

        val region = this.regionHandler[globalOriginX, globalOriginZ]
        return region.getGlobType(x, z)
    }
}
